File: lab7.3.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// Получает максимальную длину перекрытия между суффиксом first и префиксом second
func getOverlap(first, second string) int {
	maxOver := 0
	lenFirst := len(first)
	lenSecond := len(second)

	for i := 1; i <= min(lenFirst, lenSecond); i++ {
		if first[lenFirst-i:] == second[:i] {
			maxOver = i
		}
	}
	return maxOver
}

// Рекурсивно формирует все возможные цепочки слов и выбирает минимальную по длине
func makeCrossword(words []string, used []bool, curr string, minLen *int, bestRes *string, usedCount int, lastWord string) {
	if usedCount == len(words) {
		if len(curr) < *minLen {
			*minLen = len(curr)
			*bestRes = curr
		}
		return
	}

	for i := 0; i < len(words); i++ {
		if !used[i] {
			overlap := getOverlap(lastWord, words[i])

			used[i] = true
			nextWord := words[i]
			oldSize := len(curr)
			curr += nextWord[overlap:]
			makeCrossword(words, used, curr, minLen, bestRes, usedCount+1, words[i])
			curr = curr[:oldSize]
			used[i] = false
		}
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func main() {
	var c int
	fmt.Print("Введите количество слов: ")
	_, err := fmt.Scan(&c)
	if err != nil || c <= 0 {
		fmt.Println("Ошибка. Введите положительное число")
		return
	}

	words := make([]string, c)
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Введите", c, "слов:")

	for i := 0; i < c; i++ {
		if !scanner.Scan() {
			fmt.Println("Ошибка чтения ввода")
			return
		}
		words[i] = strings.TrimSpace(scanner.Text())
		if words[i] == "" {
			fmt.Println("Ошибка: пустое слово невозможно добавить")
			return
		}
	}

	minLen := int(1e9)
	bestRes := ""

	for i := 0; i < c; i++ {
		used := make([]bool, c)
		used[i] = true
		makeCrossword(words, used, words[i], &minLen, &bestRes, 1, words[i])
	}

	fmt.Println("Минимальная длина кроссворда:", minLen)
	fmt.Println("Цепочка:", bestRes)
}

File: lab7.3.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;


int getOverlap(string& first, string &second)
{
    int maxOver = 0; //длина самого длинного перекрытия
    int lenFirst = first.length();
    int lenSec = second.length();

    //перебор длин покрытия
    for(int i=1; i <= min(lenFirst, lenSec); i++)
    {   //сравниваем слово с конца, количество символов, строку с которой сравниваем, с началом, количество символов в second
        if(first.compare(lenFirst - i, i, second, 0, i) == 0) //если суффикс и префикс одинаковые
        {
            maxOver = i; //то обновляем длину перекрытия
        }
    }
    return maxOver;
}

void makeCrossword(vector<string> &words, vector<bool> &used, string curr, int& minLen, string& bestRes, int usedCount, string& lastWord)
{
    //проверка на лучший вариант перестановки
    if (usedCount == words.size()) {
        int len = curr.length();
        if (len < minLen) { 
            minLen = len;
            bestRes = curr;
        }
        return;
    }

    for (size_t i = 0; i < words.size(); ++i) {
        if (!used[i]) {//для каждого неиспользованного слова
            int overlap = getOverlap(lastWord, words[i]); //ищем перекрытие между последним и текущим словом

            used[i] = true; //метка что слово не использовалось повторно
            //добавление неперекрывающейся части слова
            string& nextWord = words[i];
            int oldSize = curr.size();
            for(int j = overlap; j<nextWord.size();j++)
            {
                curr.push_back(nextWord[j]);
            }
            makeCrossword(words, used, curr, minLen, bestRes, usedCount + 1, words[i]); //вызываем рекурсию с добавлеием слова, использованых

            curr.resize(oldSize); //откат чтобы не перемешивать изменённые строки
            used[i] = false;
        }
    }
}

int main()
{
    cout << "Введите количество слов: ";
    int c;


    if(!(cin>>c) || c <=0)
    {
        cout << "Ошибка. Введите положительное число" << endl;
        return 1;
    }

    vector<string> mp(c);
    cout << "Введите " << c << " слов:" << endl;

    for(int i=0; i <c; i++)
    {
        cin >> mp[i];
        if(mp[i].empty())
        {
            cout << "Ошибка: пустое слово невозможно добавить" << endl;
            return 1;
        }
    }
    int minLen = 1e9; //для любой доступной длины суперстроки
    string chain;

    for (size_t i = 0; i < mp.size(); ++i) {
        vector<bool> used(c, false);
        used[i] = true;
        makeCrossword(mp, used, mp[i], minLen, chain, 1, mp[i]);
    }

    cout << "Минимальная длина кроссворда: " << minLen << endl;
    cout << "Цепочка: " << chain << endl;
}